util/src/main/java/io/kubernetes/client/util/KubeConfig.java

```diff
@@ -229,6 +244,99 @@ public String getAccessToken() {
+      log.error("Unrecognized user.exec.apiVersion: {}", apiVersion);
+      return null;
+    }
+    String command = (String) execMap.get("command");
+    JsonElement root = runExec(command, (List) execMap.get("args"), (List) execMap.get("env"));
+    if (root == null) {
+      return null;
+    }
+    if (!"ExecCredential".equals(root.getAsJsonObject().get("kind").getAsString())) {
+      log.error("Unrecognized kind in response");
+      return null;
+    }
+    if (!apiVersion.equals(root.getAsJsonObject().get("apiVersion").getAsString())) {
+      log.error("Mismatched apiVersion in response");
+      return null;
+    }
+    JsonObject status = root.getAsJsonObject().get("status").getAsJsonObject();
+    JsonElement token = status.get("token");
+    if (token == null) {
+      // TODO handle clientCertificateData/clientKeyData
+      // (KubeconfigAuthentication is not yet set up for that to be dynamic)
+      log.warn("No token produced by {}", command);
+      return null;
+    }
+    log.debug("Obtained a token from {}", command);
+    return token.getAsString();
+    // TODO cache tokens between calls, up to .status.expirationTimestamp
+    // TODO a 401 is supposed to force a refresh,
+    // but KubeconfigAuthentication hardcodes AccessTokenAuthentication which does not support that
+    // and anyway ClientBuilder only calls Authenticator.provide once per ApiClient;
+    // we would need to do it on every request
+  }
+
+  private JsonElement runExec(String command, List<String> args, List<Map<String, String>> env) {
+    List<String> argv = new ArrayList<>();
+    if (command.startsWith("./")) {
+      // TODO spec is unclear on what should be treated as a “relative command path”
+      File resolvedCommand = new File(file.getParentFile(), command);
+      if (!resolvedCommand.isFile()) {
+        log.error("No such file: {}", resolvedCommand);
+        return null;
+      }
+      argv.add(resolvedCommand.getAbsolutePath());
+    } else {
+      argv.add(command);
+    }
+    if (args != null) {
+      argv.addAll(args);
+    }
+    ProcessBuilder pb = new ProcessBuilder(argv);
+    if (env != null) {
+      for (Map<String, String> entry : env) {
+        pb.environment().put(entry.get("name"), entry.get("value"));
+      }
+    }
+    pb.redirectError(ProcessBuilder.Redirect.INHERIT);
+    try {
+      Process proc = pb.start();
+      JsonElement root;
+      try (InputStream is = proc.getInputStream();
+          Reader r = new InputStreamReader(is, StandardCharsets.UTF_8)) {
+        root = new JsonParser().parse(r);
+      } catch (JsonParseException x) {
+        log.error("Failed to parse output of " + command, x);
+        return null;
+      }
+      int r = proc.waitFor();
+      if (r != 0) {
+        log.error("{} failed with exit code {}", command, r);
+        return null;
+      }
+      return root;
+    } catch (IOException | InterruptedException x) {
```

Comment:
why swallow the exception here? wouldn't you rather throw?

Reply:
It is not swallowed—it is sent to the log.

As to whether it is appropriate to throw exceptions out of `getAccessToken` (presumably wrapped in `RuntimeException`?) as opposed to logging and proceeding with other authentication modes and ultimately returning null (unauthenticated), I have no strong opinion. I was following the example of existing code which proceeds despite errors (`Failed to read token file` for example).

On the other hand `AzureActiveDirectoryAuthenticator.refresh` appears to throw an undeclared exception out of `getAccessToken`. `KubeconfigAuthentication.<init>` is permitted to throw `IOException` so perhaps `getAccessToken` should as well? That would mean reworking some existing code which catches and logs exceptions, and adding `IOException` to the signature of `Authenticator.refresh`.

Reply:
ok, sounds like it's messy, let's take it as a follow-on cleanup exercise...

Reply:
Noted in PR description.